canvas.go

```diff
@@ -49,22 +56,42 @@ type line struct {
 }
 
 // definition creates any necessary defs that can be used later in the SVG.
-func (s *service) definition(canvas *svg.SVG) {
+func (s *service) definition(canvas *svg.SVG, iconsRendered map[string]bool, iconIds map[string]string) error {
+	if s.iconSrc != "" && !iconsRendered[s.charmPath] {
```

Comment:
Perhaps return early to keep indentation small (I think the
reversed condition is a little easier to understand too):

    if s.iconSrc == "" || iconsRendered[s.charmPath] {
        return nil
    }